x/mint/types/minter.go

```diff
@@ -0,0 +1,76 @@
+package types
+
+import (
+	"fmt"
+	"time"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// NewMinter returns a new Minter object.
+func NewMinter(inflationRate sdk.Dec, annualProvisions sdk.Dec, genesisTime *time.Time, bondDenom string) Minter {
+	return Minter{
+		InflationRate:    inflationRate,
+		AnnualProvisions: annualProvisions,
+		GenesisTime:      genesisTime,
+		BondDenom:        bondDenom,
+	}
+}
+
+// DefaultMinter returns a Minter object with default values.
+func DefaultMinter() Minter {
+	unixEpoch := time.Unix(0, 0).UTC()
+	return NewMinter(initalInflationRate, sdk.NewDec(0), &unixEpoch, sdk.DefaultBondDenom)
+}
+
+// Validate returns an error if the minter is invalid.
+func (m Minter) Validate() error {
+	if m.InflationRate.IsNegative() {
+		return fmt.Errorf("inflation rate %v should be positive", m.InflationRate.String())
+	}
+	if m.AnnualProvisions.IsNegative() {
+		return fmt.Errorf("annual provisions %v should be positive", m.AnnualProvisions.String())
+	}
+	if m.BondDenom == "" {
+		return fmt.Errorf("bond denom should not be empty string")
+	}
+	return nil
+}
+
+// CalculateInflationRate returns the inflation rate for the current year depending on
+// the current block height in context. The inflation rate is expected to
+// decrease every year according to the schedule specified in the README.
+func (m Minter) CalculateInflationRate(ctx sdk.Context) sdk.Dec {
+	years := yearsSinceGenesis(*m.GenesisTime, ctx.BlockTime())
+	inflationRate := initalInflationRate.Mul(sdk.OneDec().Sub(disinflationRate).Power(years))
+
+	if inflationRate.LT(targetInflationRate) {
+		return targetInflationRate
+	}
+	return inflationRate
+}
+
+// CalculateAnnualProvisions returns the total number of tokens that should be
+// minted due to inflation for the current year.
+func (m Minter) CalculateAnnualProvisions(totalSupply math.Int) sdk.Dec {
+	return m.InflationRate.MulInt(totalSupply)
+}
+
+// CalculateBlockProvision returns the total number of coins that should be
+// minted due to inflation for the current block.
+func (m Minter) CalculateBlockProvision(current time.Time, previous time.Time) sdk.Coin {
+	timeElapsed := current.Sub(previous).Seconds()
+	portionOfYear := sdk.NewDec(int64(timeElapsed)).Quo(sdk.NewDec(int64(SecondsPerYear)))
```

Comment:
note that time elapsed is a `float64` and when we convert it to an int64 here, we're round down significantly based on the actual amount of inflation. If a block is 15.9 seconds, we're only giving out 15s worth of inflation, and thus stiffing the community their divine remuneration (:trollface:). Since we have so many blocks per, this adds up pretty quick

after 40 3 second blocks the community pool balance is:

| Using Time Unit       | Value                           |
|-----------------------|---------------------------------|
| seconds        | 143456203394.140000000000000000utia |
| milliseconds   | 173977278048.860000000000000000utia |
| nanoseconds | 174005731475.800000000000000000utia |


We already have nano second precision from the header time, so I actually think we should use that. This way, we get almost exactly the desired amount of inflation per year.

Reply:
Great catch! Resolved by using nanoseconds